--- ../untouched/gnatsweb-2.7beta/gnatsweb.pl	Sun Jan 16 01:46:01 2000
+++ gnatsweb.pl	Tue Jun 26 21:49:11 2001
@@ -1921,7 +1921,7 @@ sub delete_stored_query
 sub send_html
 {
   my $file = shift;
-  open(HTML, "$file") || die "Can't open $file: $!";
+  open(HTML, "<$file") || die "Can't open $file: $!";
   undef $/; # slurp file whole
   my $html = <HTML>;
   close(HTML);
@@ -1935,7 +1935,7 @@ sub send_html
 
 sub help_page
 {
-  my $html_file = $q->param('help_file') || 'gnatsweb.html';
+  my $html_file = 'gnatsweb.html';
   my $page      = $q->param('help_title') || 'Help';
   my $heading   = $page;
   page_start_html($page);
